autode/opt/coordinates/autodiff.py

```diff
@@ -0,0 +1,535 @@
+                + np.outer(other._first_der, self._first_der)
+                + other._val * self._second_der
+            )
+            return VectorHyperDual(val, self._symbols, first_der, second_der)
+        else:
+            raise TypeError("Unknown type for multiplication")
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    def __truediv__(self, other):
+        """True division, defined by multiplicative inverse"""
+        return self.__mul__(DifferentiableMath.pow(other, -1))
+
+    def __rtruediv__(self, other):
+        """Reverse true division"""
+        return DifferentiableMath.pow(self, -1).__mul__(other)
+
+    def __pow__(self, power, modulo=None) -> "VectorHyperDual":
+        if modulo is not None:
+            raise NotImplementedError("Modulo inverse is not implemented")
+
+        return DifferentiableMath.pow(self, power)
+
+    def __rpow__(self, other):
+        return DifferentiableMath.pow(other, self)
+
+    @staticmethod
+    def apply_operation(
+        num: Union["VectorHyperDual", numeric_type],
+        operator: Callable[[float], float],
+        operator_first_deriv: Callable[[float], float],
+        operator_second_deriv: Callable[[float], float],
+    ):
+        """
+        Perform an operation on the hyperdual (i.e. apply a scalar function)
+
+        Args:
+            num: Number that is hyper-dual (or float/int)
+            operator: Function that returns the value (result) of operation
+            operator_first_deriv: Should return first derivative of operation
+            operator_second_deriv: Should return second derivative of operation
+
+        Returns:
+            (VectorHyperDual|float): The result
+        """
+        # pass through numeric types
+        if isinstance(num, numeric):
+            return operator(float(num))
+
+        val = operator(num._val)
+
+        if num._order == 0:
+            return VectorHyperDual(val, num._symbols)
+
+        assert num._first_der is not None
+        f_dash_x0 = operator_first_deriv(num._val)
+        first_der = num._first_der * f_dash_x0
+
+        if num._order == 1:
+            return VectorHyperDual(val, num._symbols, first_der)
+
+        assert num._second_der is not None
+        second_der = np.outer(num._first_der, num._first_der) * (
+            operator_second_deriv(num._val)
+        )
+        second_der += num._second_der * f_dash_x0
+        return VectorHyperDual(val, num._symbols, first_der, second_der)
+
+
+class DifferentiableMath:
+    """
+    Class defining math functions that can be used on
+    hyper dual numbers (i.e. differentiable functions),
+    as well as standard numeric types (float and int)
+    """
+
+    @staticmethod
+    def sqrt(num: Union[VectorHyperDual, numeric_type]):
+        """Calculate the square root of a hyperdual number"""
+
+        if isinstance(num, numeric):
+            assert num > 0
+        else:
+            assert num.value > 0
```

Comment:
@t-young31 I considered that, but it is harder to detect errors in that case, since `math` functions implicitly convert to float. So I made the differentiable numbers only compatible with the DifferentiableMath functions.